include/cliffordsynthesis/CliffordSynthesizer.hpp

```diff
@@ -0,0 +1,289 @@
+            const LogicMatrix3D& g_s, const LogicMatrix3D& g_c) const;
+    void make_fidelity_optimizer(
+            int                                                        timesteps,
+            const std::set<std::pair<unsigned short, unsigned short>>& reducedCM,
+            const std::vector<unsigned short>& qubitChoice, std::unique_ptr<LogicBlock>& lb,
+            const LogicMatrix& x, const LogicMatrix& z, const LogicVector& r,
+            const LogicMatrix3D& g_s, const LogicMatrix3D& g_c) const;
+
+    void        runMinimizer(int timesteps, const CouplingMap& reducedCM,
+                             const std::vector<unsigned short>& qubitChoice);
+    void        runStartLow(int timesteps, const CouplingMap& reducedCM,
+                            const std::vector<unsigned short>& qubitChoice);
+    void        runStartHigh(int timesteps, const CouplingMap& reducedCM,
+                             const std::vector<unsigned short>& qubitChoice);
+    void        runMinMax(int timesteps, const CouplingMap& reducedCM,
+                          const std::vector<unsigned short>& qubitChoice);
+    void        runSplitIter(const CouplingMap&                 reducedCM,
+                             const std::vector<unsigned short>& qubitChoice);
+    static void runSplinter(int i, unsigned int circ_split, unsigned int split,
+                            const CouplingMap&                 reducedCM,
+                            const std::vector<unsigned short>& qubitChoice,
+                            qc::QuantumComputation&            circuit,
+                            CliffordOptimizationResults* r, CliffordOptimizer* opt);
+    void        updateResults(CliffordOptimizationResults& r);
+
+    static void assertTableau(const Tableau& tableau, std::unique_ptr<LogicBlock>& lb,
+                              const LogicMatrix& x, const LogicMatrix& z,
+                              const LogicVector& r, int nqubits, int position);
+
+    static void makeSingleGateConstraints(
+            std::unique_ptr<LogicBlock>& lb, const LogicMatrix& x, const LogicMatrix& z,
+            const LogicVector& r, int nqubits, int timesteps,
+            const std::set<std::pair<unsigned short, unsigned short>>& reducedCM,
+            const std::vector<unsigned short>& qubitChoice, const LogicMatrix3D& g_s,
+            const LogicMatrix3D& g_c);
+    static void makeMultipleGateConstraints(
+            std::unique_ptr<LogicBlock>& lb, const LogicMatrix& x, const LogicMatrix& z,
+            const LogicVector& r, int nqubits, int timesteps,
+            const std::set<std::pair<unsigned short, unsigned short>>& reducedCM,
+            const std::vector<unsigned short>& qubitChoice, const LogicMatrix3D& g_s,
+            const LogicMatrix3D& g_c);
+};
+
+class Gates {
+public:
+    enum GATES {
+        NOP,
+        H,
+        S,
+        X,
+        Y,
+        Z,
+        Sdag,
+        CX,
+        CY,
+        CZ,
+    };
```

Comment:
To me, this class seems rather redundant when looking at the OpType enum. The collections of supported gates could just as well be collections of OpTypes. The `NOP` can just be replaced by the identity gate.

Reply:
The main problem I had with using OpTypes was that it was harder to discern controlled gates and single qubit gates, so this Class was the easiest solution

Reply:
Hm. I see your point. Not totally happy with the solution, but I also don't have a good alternative in mind. Let's keep this for now.

Just as a side note: this list is still missing the SWAP gate.

Reply:
SWAP is not a Gate the Synthesizer produces, which is why its not included.